proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,160 @@
+These rules also depend on the target IR, SPIR-V or DXIL.
+
+RETs for TypedBuffer variants may include:
+* basic types: 
+  * 16- and 32-bit int and uint
+  * half and float
+* vectors and matrices 
+  * containing 4 elements or fewer
+  * total size may not exceed 128 bits
+* user defined types (structs / classes), as long as:
+  * all fields in the struct have the same type
+  * there are at most 4 sub elements
+  * total size may not exceed 128 bits
+
+RETs for RawBuffer variants are much less constrained:
+* it must be a complete type
+* cannot contain a handle or resource type
+
+Resource types are never allowed as RETs (i.e., `RWBuffer<int>` as an RET).
+Texture resources conform to the rules for TypedBuffers.
+If the target is SPIR-V, then only the set of rules for RawBuffers apply. TypedBuffers
+like `StructuredBuffer` may have RETs that exceed 128 bits, for example, if the target 
+IR is SPIR-V. The TypedBuffer rules above are only enforced when the IR target is DXIL.
+
+If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a 
+valid RET, there should be infrastructure to reject this RET and emit a message 
+explaining why it was rejected as an RET.
+
+## Motivation
+Currently, there is an allow list of valid RETs. It must be modified with respect 
+to this spec. Anything that is not an int, uint, nor a floating-point type, or vectors 
+or matrices containing the aforementioned types, will be rejected. The allow list isn't
+broad enough, because it doesn't include the case where the RET is user-defined. 
+Ideally, a user should be able to determine how any user-defined structure is invalid 
+as an RET. Some system should be in place to more completely enforce the rules for 
+valid and invalid RETs, as well as provide useful information on why they are invalid.
+
+For example, when targeting DXIL IR, `RWBuffer<double4> b : register(u4);` will emit
+an error in DXC, but will not in clang-dxc, despite the fact that `double4` is an 
+invalid RET for TypedBuffers.
+
+## Proposed solution
+
+The proposed solution is to use some type_traits defined in the std library, create
+some custom builtins to use as type_traits, and join them together to define a 
+set of conceptual constraints for any RET that is used. These conceptual constraints
+will be applied to every TypedBuffer resource type that is defined, so that all
+TypedBuffer HLSL resources have the same rules about which RETs are valid. 
+Validation will occur upon resource type instantiation. Additionally, certain 
+resource types are RawBuffer variants, such as `StructuredBuffer`. These
+resource types will have a different set of type-traits applied, which will
+loosen constraints on viable RETs.
+
+## Detailed design
+
+In `clang\lib\Sema\HLSLExternalSemaSource.cpp`, `RWBuffer` is defined, along with 
+`RasterizerOrderedBuffer` and `StructuredBuffer`. It is at this point that the 
+`type_traits` should be incorporated into these resource declarations. The 

```

Comment:
I don't understand what the plan is for how these type traits are going to be used?  Just having a type trait doesn't automatically make diagnostics get emitted.

Are you planning to add static_asserts to the resource class definitions?  If so, I'd really like to see some examples of how this will be done.